### PR TITLE
Modes of interaction

### PR DESCRIPTION
**Pull request by martinthomson**

### CODE REVIEW

**Comment by martinthomson** [comment]
@mnot, does this all seem reasonable?  (If you would prefer to defer reviewing until we push a draft, that's OK.)

File: draft-ietf-git-using-github.md [hunk 1]
```diff
@@ -419,6 +435,239 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+or editors record details of decisions in issue comments when closing issues as
+resolved.
+
+
+## Heavy GitHub Usage
+
+The "heavy" interaction mode differs from the "light" mode in that discussion
+relating to substantive technical matters is allowed to occur on GitHub issues.
+As a result, participants are permitted to contribute to discussion on the issue
+tracker.  In some cases, this can include making some decisions without
```
Comment (mnot) [inline draft-ietf-git-using-github.md:521]: This is going to be cause a lot of ire. I'd suggest at least parenthetically saying that those decisions are confirmed on the list. I know it's said below, but people will read this out of context (or just complain before reading onwards).

File: draft-ietf-git-using-github.md [hunk 2]
```diff
@@ -419,6 +435,239 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+
+* An `invalid` label might be used for issues that were raised in error.
+
+* A `blocked` label might indicate an issue is awaiting resolution of an
+  external process or related issue.
+
+* A `parked` label might be used to indicate issues that do not require
+  immediate working group attention.
+
+
```
Comment (mnot) [inline draft-ietf-git-using-github.md:694]: We use a `discusss` flag for http-core to identify issues to talk about F2F.

**Review by mnot — COMMENTED** [review]
Seems like a good start. I suspect all of this is going to evolve, but no harm in writing it down.

File: draft-ietf-git-using-github.md (hunk 2 above)
Comment (martinthomson) [reply draft-ietf-git-using-github.md:694]: I considered that, but I think that it just falls into the "other other"  bucket.  We use the project boards in QUIC for doing the same thing.  I wasn't even sure about including `parked` here.

**Comment by bhstark2** [comment]
I'm thinking maybe there needs to be a heavy, medium and light. This text describes medium and heavy -- not light. I'll start writing what I think is light today. The main problem is y'all are coming at this from the perspective of people with considerable GitHub expertise and comfort-level. I'm coming at it from a perspective of "I've no clue what I'm doing; it's all magic." Where there is one person like me, there are many.

**Comment by martinthomson** [comment]
Sounds good.  I am probably incapable of seeing what a truly light option looks like.

**Comment by bhstark2** [comment]
Since proposing text to extend what's in an un-merged PR is something I'm really not comfortable with doing, I'll just paste here what I was thinking about ultra-light.

## Ultra-light

There are varying levels of GitHub expertise in IETF. It cannot be assumed that ADs, chairs, authors or contributors have an existing login to GitHub or have ever used it. For some WGs, it may make sense to start with an ultra-light approach to using GitHub. This would allow draft authors to make use of basic GitHub functionality and allow WG chairs to get some initial experience (if this is the level they are at in their GitHub journey). Independent of experience or knowledge, this approach may also be useful for a WG happy with its current way of doing work, that would like to make use of GitHub as a very simple tool to enhance existing methods of collaboration.

GitHub can be used in an "ultra-light" manner by setting it up primarily for use by draft authors. They can use issues and comments to discuss the draft among themselves, and they can use it to provide WG members with a preview "Editor's Copy" to confirm on-list comments have been adequately resolved (prior to submitting to Datatracker).

All WG discussion happens on the email list.

Using GitHub in this manner is simplest if each draft has its own repository.

The repository can be set up with the authors as a team with Write access. They can decide among themselves how they want to work. All other GitHub members have Read access. The WG is informed of the existence of the repository, but told that all issues are to be discussed on the list. 

The authors can choose to make full use of the available tools for automating processes, or manually perform Datatracker submissions and GitHub pages. If the draft is written using XML, it is even possible to avoid installing any tools. The on-line version of xml2rfc can be used to check the draft, and the GitHub UI can be used to upload draft versions. For users of Microsoft Windows, it is also possible to use [Git for Windows](https://gitforwindows.org/) to interact with GitHub for for all aspects of ultra-light interaction.

File: draft-ietf-git-using-github.md [hunk 3]
```diff
@@ -428,6 +447,242 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+meetings, and other more traditional interactions.  The intent of this policy is
+to enable document and issue management using GitHub while minimizing the
+complexity of the process.
+
+The working group uses GitHub for tracking outstanding issues and all
+substantive issues require that an issue be opened to record the issue, but
+discussion of any substantial matters is always conducted on mailing lists.
+
+Under this mode, issues and pull requests can be opened by anyone, but anything
+deemed substantive MUST resolved exclusively on the mailing list.  Discussion on
```
Comment (chris-wood) [inline draft-ietf-git-using-github.md]:
```suggestion
deemed substantive MUST be resolved exclusively on the mailing list.  Discussion on
```

**Review by chris-wood — APPROVED** [review]
LGTM! I think we can land the light changes from Barbara in a separate PR.

**Comment by chris-wood** [comment]
@bhstark2 the proposed "Ultra Light" profile seems mostly the same as what Martin provides. I would remove all content regarding document editing and submission (last paragraph), as it does not pertain to the purpose of this WG document. I might also move the contents of your first paragraph to the "Typical Working Group Policies" section in Martin's change. That's where information about choosing which mode is right seems best fit. 

@martinthomson, what do you think?
@paulehoffman, please have a look.

**Comment by martinthomson** [comment]
In looking at this, I realize that there is a progressive thing at play more than we have a discrete "modes" of engagement.

At the lightest end, we have people who just throw stuff up on GitHub as some place stable to keep some files.  And that can progress to having issue tracking there for editor convenience, all the way to having the working group actively relying on the tracker.  As people get more familiar with how things work, then you see more demand and use.  I'm wondering whether we should try to shave this down further and try to describe how the progression might work, recognizing that every group will have its own sweet spot on that and that is OK.

**Comment by chris-wood** [comment]
> I'm wondering whether we should try to shave this down further and try to describe how the progression might work, recognizing that every group will have its own sweet spot on that and that is OK.

In principle, I think this is a fine idea. GitHub usage is certainly a spectrum, and defining that spectrum will be challenging. Maybe we can describe this concept loosely in the {{modes}} introduction to start?

**Comment by coopdanger** [comment]
> @bhstark2 the proposed "Ultra Light" profile seems mostly the same as what Martin provides. I would remove all content regarding document editing and submission (last paragraph), as it does not pertain to the purpose of this WG document. I might also move the contents of your first paragraph to the "Typical Working Group Policies" section in Martin's change. That's where information about choosing which mode is right seems best fit.
> 
> @martinthomson, what do you think?
> @paulehoffman, please have a look.

Perhaps the main difference between what Barbara is suggesting and what Martin wrote is the expectation about issue discussion and resolution. I think the "ultra-light" description assumes no one will be opening issues aside from document editors, in which case the bit about having to be vigilant to return issue discussion to the list doesn't really apply because the editors are just using the issue tracker for their own purposes. One further step in the progression is when WG participants start opening issues themselves, at which point someone needs to be responsible for directing discussion to the venue the WG has chosen (issues vs. mailing list). Maybe this could be reflected in the description of the spectrum.

**Comment by chris-wood** [comment]
> One further step in the progression is when WG participants start opening issues themselves, at which point someone needs to be responsible for directing discussion to the venue the WG has chosen (issues vs. mailing list). Maybe this could be reflected in the description of the spectrum.

That seems reasonable to me. I do worry of us trying to enumerate each option in the spectrum, though, as there are undoubtedly many GitHub knobs to turn!

**Comment by coopdanger** [comment]
> > One further step in the progression is when WG participants start opening issues themselves, at which point someone needs to be responsible for directing discussion to the venue the WG has chosen (issues vs. mailing list). Maybe this could be reflected in the description of the spectrum.
> 
> That seems reasonable to me. I do worry of us trying to enumerate each option in the spectrum, though, as there are undoubtedly many GitHub knobs to turn!

Agreed, just focusing on the key differences in how a WG can engage with GitHub is sufficient. I do think this question of who is expected to file issues is one of them.

File: draft-ietf-git-using-github.md [hunk 4]
```diff
@@ -428,6 +447,242 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+mailing list, along lines similar to the design team process (see Section 6.5 of
+{{RFC2418}}).
+
+As a more involved process, adopting this mode can require changes in policies
+as documents become more mature.  It is possible to use a different process for
+each active document in the same Working Group.  Working Group chairs MUST
+ensure that the process that applies to different documents is clear, which
+includes both maintaining clear documentation about policies that relate to each
+document and providing suitable advance notice of changes to policies on the
+Working Group mailing list.
```
Comment (josephlhall) [inline draft-ietf-git-using-github.md]: the above paragraph lost me... it's just a bit circuituous. I'm not sure it's needed, but if so, I'd suggest making it very simple ("It is possible to use a different process for each active document in the same Working Group.").

File: draft-ietf-git-using-github.md [hunk 5]
```diff
@@ -428,6 +447,242 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+
+Gaining Working Group consensus about the resolution of issues can be done in
+the abstract, with editors being permitted to capture the outcome of discussions
+as they see fit.
+
+More mature documents require not only consensus, but consensus about specific
+text.  All substantive changes to documents that have passed WGLC MUST be
+proposed as pull requests, discussed on the mailing list, and have chairs
+explicitly confirm consensus.  Chairs MAY institute this stricter process prior
+to WGLC.
```
Comment (josephlhall) [inline draft-ietf-git-using-github.md]: Doesn't it seem sufficient to have chairs confirm consensus and not necessarily require pull requests? (E.g., have an editor make changes directly to HEAD... or is there a rationale for PRs that I can't read between the lines to see here?). I'm just saying making that a MUST means lots of PRs that may not have to have been? Or maybe post-WGLC we definitely want PRs?

File: draft-ietf-git-using-github.md [hunk 6]
```diff
@@ -428,6 +447,242 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+
+An issue labeled as `editorial` has no substantive effect on a document, except
+to the extent that addressing the issue might make understanding the
+specification easier.  Resolution of `editorial` issues can be left to the
+discretion of editors.
+
+This distinction can be applied to all types of document.  For instance, a
+`design` issue for an Informational document might be raised to discuss
+important concepts.
+
```
Comment (josephlhall) [inline draft-ietf-git-using-github.md:645]: Hmmm, "usually affecting an implementation or interop" (above) seems pretty different from "discussing important concepts". Maybe the right distinction for an informational document is a substantive change is `design` (something that changes the meaning of the document). I may be over-thinking this.

Comment (martinthomson) [reply draft-ietf-git-using-github.md:645]: Over-thinking or not, it's still quite helpful.

**Comment by martinthomson** [comment]
Thanks for all demonstrating how hard it is to avoid discussing things on pull requests :)

Seriously though, that was very useful feedback.  I've tweaked this section to move it to three processes, with the third being considerably more complex.  I hope that is clearer now.

**Comment by chris-wood** [comment]
@mnot can you please have a look at the latest changes? I think this is ready to go.

**Comment by chris-wood** [comment]
Also requesting review from @coopdanger, @britram, @rsalz, @MikeBishop, and @ekr, if you have time. :-)

**Comment by richsalz** [comment]
This looks good to me.  "Modes" is an excellent way of talking/thinking about the styles of interaction, and much less "opinionated" (?) than light/heavy/xxl.

File: draft-ietf-git-using-github.md (hunk 6 above)
Comment (MikeBishop) [reply draft-ietf-git-using-github.md:645]: For a protocol spec, anything that breaks compat is design.  For an Informational spec, anything that changes what a reader might choose to do with the information is a design change.  Discussing new topics or changing position on a topic would therefore be a design change, in my mind.  But I agree, there's less of a bright line in a document that doesn't eventually lead to bits on the wire.

File: draft-ietf-git-using-github.md [hunk 7]
```diff
@@ -428,6 +447,251 @@ A working group that uses GitHub MAY provide either facility at the request of
...
+a future version of or extension to a protocol, particularly where a resolution
+is made to take no action.
+
+
+### Component Labelling
+
+Repositories with multiple interrelated documents or a complex document with
+multiple logical components might benefit from labels that identify different
+aspects of the work.  The choice of appropriate labels for components will
+depend on the structure of specific documents.
```
Comment (MikeBishop) [inline draft-ietf-git-using-github.md:678]: We might want to suggest that labeling of this sort be at the editors' discretion.

**Review by MikeBishop — COMMENTED** [review]
Overall, I think this is a good way to frame it.

File: draft-ietf-git-using-github.md (hunk 7 above)
Comment (chris-wood) [reply draft-ietf-git-using-github.md:678]: @MikeBishop, could you file an issue or (better) create a PR?

**Pull request merged**